specs/core/0_beacon-chain.md

```diff
@@ -999,130 +990,161 @@ The output of `get_block_hash` should not change, except that it will no longer
 
 ```python
 def update_ancestor_hashes(parent_ancestor_hashes: List[Hash32],
-                           parent_slot_number: int,
+                           parent_slot: int,
                            parent_hash: Hash32) -> List[Hash32]:
     new_ancestor_hashes = copy.copy(parent_ancestor_hashes)
     for i in range(32):
-        if parent_slot_number % 2**i == 0:
+        if parent_slot % 2**i == 0:
             new_ancestor_hashes[i] = parent_hash
     return new_ancestor_hashes
 ```
 
 ### Verify attestations
 
-Verify that there are at most `MAX_ATTESTATION_COUNT` `AttestationRecord` objects.
+* Verify that `len(block.attestations) <= MAX_ATTESTATIONS_PER_BLOCK`.
 
-For each `AttestationRecord` object `obj`:
+For each `attestation` in `block.attestations`:
 
-* Verify that `obj.data.slot <= block.slot - MIN_ATTESTATION_INCLUSION_DELAY` and `obj.data.slot >= max(parent.slot - CYCLE_LENGTH + 1, 0)`.
-* Verify that `obj.data.justified_slot` is equal to `justification_source if obj.data.slot >= state.last_state_recalculation_slot else prev_cycle_justification_source`
-* Verify that `obj.data.justified_block_hash` is equal to `get_block_hash(state, block, obj.data.justified_slot)`.
-* Verify that either `obj.data.last_crosslink_hash` or `obj.data.shard_block_hash` equals `state.crosslinks[shard].shard_block_hash`.
+* Verify that `attestation.data.slot <= block.slot - MIN_ATTESTATION_INCLUSION_DELAY`.
+* Verify that `attestation.data.slot >= max(parent.slot - EPOCH_LENGTH + 1, 0)`.
+* Verify that `attestation.data.justified_slot` is equal to `state.justified_slot if attestation.data.slot >= state.last_state_recalculation_slot else state.previous_justified_slot`.
+* Verify that `attestation.data.justified_block_hash` is equal to `get_block_hash(state, block, attestation.data.justified_slot)`.
+* Verify that either `attestation.data.last_crosslink_hash` or `attestation.data.shard_block_hash` equals `state.crosslinks[shard].shard_block_hash`.
 * `aggregate_sig` verification:
-    * Let `participants = get_attestation_participants(state, obj.data, obj.attester_bitfield)`
-    * Let `group_public_key = BLSAddPubkeys([state.validators[v].pubkey for v in participants])`
-    * Check `BLSVerify(pubkey=group_public_key, msg=obj.data, sig=aggregate_sig, domain=get_domain(state.fork_data, slot, DOMAIN_ATTESTATION))`.
-* [TO BE REMOVED IN PHASE 1] Verify that `shard_block_hash == bytes([0] * 32)`.
-* Append `ProcessedAttestation(data=obj.data, attester_bitfield=obj.attester_bitfield, poc_bitfield=obj.poc_bitfield, slot_included=block.slot)` to `state.pending_attestations`.
+    * Let `participants = get_attestation_participants(state, attestation.data, attestation.participation_bitfield)`.
+    * Let `group_public_key = BLSAddPubkeys([state.validator_registry[v].pubkey for v in participants])`.
+    * Verify that `BLSVerify(pubkey=group_public_key, msg=attestation.data, sig=aggregate_sig, domain=get_domain(state.fork_data, slot, DOMAIN_ATTESTATION))`.
```

Comment:
You want `msg=SSZTreeHash(attestation.data) + bytes1(0)`.

Reply:
Also, I would add a one line, verify `attestation.custody_bitfield == bytes([0] * len(attestation.participation_bitfield))`.